crates/worker/src/partition/shuffle.rs

```diff
@@ -232,26 +411,175 @@ where
-                    let _ = truncation_tx.try_send(OutboxTruncation::new(shuffled_message_index));
-                },
-                _ = cancellation_watcher() => {
-                    break;
-                }
+            // this is just a hint which we can drop
+            let _ = truncation_tx.try_send(OutboxTruncation::new(shuffled_message_index));
+        }
+    }
+}
+
+mod state_machine {
+    use std::cmp::Ordering;
+
+    use tokio_util::sync::ReusableBoxFuture;
+
+    use restate_core::network::TransportConnect;
+    use restate_ingestion_client::{IngestFuture, IngestionClient, RecordCommit};
+    use restate_storage_api::outbox_table::OutboxMessage;
+    use restate_types::{identifiers::WithPartitionKey, message::MessageIndex};
+    use restate_wal_protocol::Envelope;
+
+    use crate::partition::shuffle::{
+        NewOutboxMessage, OutboxReaderError, ShuffleMetadata, wrap_outbox_message_in_envelope,
+    };
+
+    type ReadFuture<OutboxReader> = ReusableBoxFuture<
+        'static,
+        (
+            Result<Option<(MessageIndex, OutboxMessage)>, OutboxReaderError>,
+            OutboxReader,
+        ),
+    >;
+
+    async fn get_next_message<OutboxReader: super::OutboxReader>(
+        mut outbox_reader: OutboxReader,
+        next_sequence_number: MessageIndex,
+    ) -> (
+        Result<Option<(MessageIndex, OutboxMessage)>, OutboxReaderError>,
+        OutboxReader,
+    ) {
+        let result = outbox_reader.get_next_message(next_sequence_number).await;
+        (result, outbox_reader)
+    }
+
+    enum State {
+        Idle,
+        ReadingOutbox,
+        Ingesting { ingest: IngestFuture, sn: u64 },
+    }
+
+    pub struct StateMachine<T, R> {
+        metadata: ShuffleMetadata,
+        ingestion: IngestionClient<T, Envelope>,
+        hint_rx: async_channel::Receiver<NewOutboxMessage>,
+        reader: Option<R>,
+        read_fut: ReadFuture<R>,
+        next_sequence_number: MessageIndex,
+        state: State,
+    }
+
+    impl<T, R> StateMachine<T, R>
+    where
+        T: TransportConnect,
+        R: super::OutboxReader + Send + Sync + 'static,
+    {
+        pub fn new(
+            metadata: ShuffleMetadata,
+            ingestion: IngestionClient<T, Envelope>,
+            reader: R,
+            hint_rx: async_channel::Receiver<NewOutboxMessage>,
+        ) -> Self {
+            Self {
+                metadata,
+                ingestion,
+                hint_rx,
+                reader: None,
+                read_fut: ReusableBoxFuture::new(get_next_message(reader, 0)),
+                next_sequence_number: 0,
+                state: State::ReadingOutbox,
             }
         }
 
-        debug!("Stopping shuffle");
+        pub async fn shuffle_next_message(&mut self) -> anyhow::Result<RecordCommit<u64>> {
+            loop {
+                match &mut self.state {
+                    State::Idle => {
+                        loop {
```

Comment:
Is this inner loop needed or could it be removed and let the outer loop handle things?